lib/pbio/drv/usb/usb_stm32.c

```diff
@@ -164,11 +304,65 @@ PROCESS_THREAD(pbdrv_usb_process, ev, data) {
 
     PROCESS_BEGIN();
 
+    // Prepare to receive the first packet
+    USBD_Pybricks_ReceivePacket(&husbd);
+
+    etimer_set(&timer, 500);
+
     for (;;) {
         PROCESS_WAIT_EVENT();
 
         if (bcd_busy) {
             bcd_busy = PT_SCHEDULE(pbdrv_usb_stm32_bcd_detect(&bcd_pt));
+
+            if (bcd_busy) {
+                continue;
+            }
+        }
+
+        if (pbsys_status_test(PBIO_PYBRICKS_STATUS_SHUTDOWN_REQUEST) ||
+            pbsys_status_test(PBIO_PYBRICKS_STATUS_SHUTDOWN)) {
+            if (pbio_oneshot(true, &pwrdn_oneshot)) {
+                USBD_Stop(&husbd);
+            }
+
+            continue;
+        }
+
+        if (usb_in_sz) {
+            if ((usb_in_sz >= UUID_SZ) &&
+                pbio_uuid128_le_compare(usb_in_buf, pbio_pybricks_command_event_char_uuid)) {
+                pbsys_command(usb_in_buf + UUID_SZ, usb_in_sz - UUID_SZ);
+            }
+
+            usb_in_sz = 0;
+
+            // Prepare to receive the next packet
+            USBD_Pybricks_ReceivePacket(&husbd);
+        }
+
+        if (usb_out_sz) {
+            continue;
+        }
+
+        new_status_flags = pbsys_status_get_flags();
+
+        if (usb_stdout_sz) {
+            memcpy(usb_out_buf, usb_stdout_buf, usb_stdout_sz);
+            usb_out_sz = usb_stdout_sz;
+            usb_stdout_sz = 0;
+        } else if ((new_status_flags != prev_status_flags) || etimer_expired(&timer)) {
+            pbio_uuid128_le_copy(usb_out_buf, pbio_pybricks_command_event_char_uuid);
+            pbio_pybricks_event_status_report(&usb_out_buf[UUID_SZ], new_status_flags);
```

Comment:
we should be using the return value of `pbio_pybricks_event_status_report()` to get the size instead of hard-coding it below.